File: 034.Find_First_and_Last_Position_of_Element_in_Sorted_Array/Solution.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
    int start = 0;
    int end = nums.length - 1;
    int[] ans = { -1, -1 };
    if (nums.length == 0) {
        return ans;
    }
    while (start <= end) {
        int mid = (start + end) / 2;
        if (target == nums[mid]) {
            end = mid - 1;
        } else if (target < nums[mid]) {
            end = mid - 1;
        } else {
            start = mid + 1;
        }
    }
    //考虑 tartget 是否存在，判断我们要找的值是否等于 target 并且是否越界
    if (start == nums.length || nums[ start ] != target) {
        return ans;
    } else {
        ans[0] = start;
    }
    ans[0] = start;
    start = 0;
    end = nums.length - 1;
    while (start <= end) {
        int mid = (start + end) / 2;
        if (target == nums[mid]) {
            start = mid + 1;
        } else if (target < nums[mid]) {
            end = mid - 1;
        } else {
            start = mid + 1;
        }
    }
    ans[1] = end;
    return ans;
}
}

File: 034.Find_First_and_Last_Position_of_Element_in_Sorted_Array/Solution.py
class Solution:
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        length = len(nums)
        if length == 0:
            return [-1,-1]
        if nums[0] <= target and nums[-1] >= target:
            left,right = 0,length-1

            while left <= right:
                mid = (left+right) // 2

                if nums[mid] == target:
                    right = left = mid

                    while left-1 >= 0 and nums[left-1] == target:
                        left -= 1
                    while right+1 <= length-1 and nums[right+1] == target:
                        right += 1
                    return [left,right]

                elif nums[mid] < target:
                    left = mid + 1
                else:
                    right = mid - 1

        return [-1,-1]